### ISSUE TITLE
More robust pattern matching

### ISSUE THREAD

**Issue by kxfang** [airbnb/DeepLinkDispatch#144]
Our app supports all of the following:

http://example.com/foo
https://example.com/foo
http://www.example.com/foo
https://www.example.com/foo

It'd be nice to make the URL pattern matching more robust such that there's no need to have to write all those variants out everywhere. Looking at the implementation in DeepLinkEntry this seems feasible, but we'd need to agree on a syntax that works well with the existing pattern matching. My intuition is that supporting regex fully might lead to a lot of weird edge cases, so maybe we can support something simple? Happy to help/contribute to this if there's interest.

**Comment by felipecsl** [comment]
Do you have something in mind on the API for this?

**Comment by kxfang** [comment]
Spent more time thinking about this, maybe robust regex/pattern matching is the wrong way to go. Perhaps a better problem to solve is the problem that there is no way to declare scheme + host separately from the path. Currently there's a bit of ugly repetition in @DeepLink since you always need to specify "example://<path>" or "http://example.com/<path>". 

My proposal is then to support client-defined annotations that declare properties about the deep link, so you'd define your specs like this:
```
@DeepLinkSpec(scheme = "example")
public @interface @AppDeepLink {
}

@DeepLinkSpec(
    scheme = { "http", "https" }, 
    host = { "example.com", "www.example.com" })
public @interface @WebDeepLink {
}
```
And then the actual handlers would look like this:
```
@AppDeepLink({"/foo/bar", "/baz/baz2"})
public static Intent intentForAppDeepLinkMethod(Context context) {
  // return Intent to handle "example://foo/bar" and "example://baz/baz2"
}

@WebDeepLink("/foo/bar")
public static Intent intentForWebDeepLinkMethod(Context context) {
  // return Intent to handle "http://example.com/foo/bar", "https://example.com/foo/bar",
  // "http://www.example.com/foo/bar" and "https://www.example.com/foo/bar"
}
```
We can keep the @DeepLink annotation around with legacy behavior for backwards compatibility.

What do you think?

**Comment by felipecsl** [comment]
This is an interesting approach and it would definitely clean up the ugly repetition in the `@DeepLink` annotations, as you mentioned.
Couple of question:

1. Can you actually declare an `intent-filter` with multiple `scheme`s for the same Activity? I've never tried it, but seems possible;
1. In your example above for `@WebDeepLink`, does that mean that would match any permutation of URIs with the 2 schemes and 2 hosts declared? is that intentional?

**Comment by kxfang** [comment]
1. Yep, definitely possible. We're doing it on our project.
2. Yes in my original example I did intend for it to match any permutation of URIs with the 2 schemes and 2 hosts declared. I can see that might be contentious though, and I would be perfectly happy if it was something like the following, which is a little more verbose but less confusing:
```
@DeepLinkSpec(
    prefix = { "http://example.com", 
                "https://example.com", 
                "http://www.example.com", 
                "https://www.example.com",  }
public @interface @WebDeepLink {
}
```

**Comment by felipecsl** [comment]
Yeah I feel like the `prefix` option would be simpler/easier to understand. That seems fine by me 😃

**Comment by kxfang** [comment]
Cool let me look into this

**Comment by felipecsl** [comment]
Awesome!!

**Comment by felipecsl** [comment]
Closed via #147

**Comment by hetang** [comment]
@felipecsl How to support multiple subdomain in prefix? I do not see any documentation related this. Can you please help me here?
